hudi-hadoop-common/src/main/java/org/apache/parquet/avro/HoodieAvroReadSupport.java

```diff
@@ -62,6 +58,23 @@ public ReadContext init(Configuration configuration, Map<String, String> keyValu
     return new ReadContext(requestedSchema, readContext.getReadSupportMetadata());
   }
 
+  /**
+   * Here we want set config with which file has been written.
+   * Even though user may have overwritten {@link AvroWriteSupport.WRITE_OLD_LIST_STRUCTURE},
+   * it's only applicable to how to produce new files(here is a read path).
+   * Later the config value {@link AvroWriteSupport.WRITE_OLD_LIST_STRUCTURE} will still be used
+   * to write new file according to the user preferences.
+   **/
+  private void adjustConfToReadWithFileProduceMode(Boolean isLegacyModeWrittenFile, Configuration configuration) {
+    if (isLegacyModeWrittenFile) {
+      configuration.set(AvroWriteSupport.WRITE_OLD_LIST_STRUCTURE,
+          "true", "support reading avro from legacy map/list in parquet file");
+    } else {
+      configuration.set(AvroWriteSupport.WRITE_OLD_LIST_STRUCTURE,
+          "false", "support reading avro from non-legacy map/list in parquet file");
+    }
```

Comment:
I personally think this is an OK fix. An alternative may be adding something like `HoodieAvroSchemaConverter` to rewrite schema within Hudi. But it would be too much for such a corner case and could become hard to maintain in the future when Parquet changes their `AvroSchemaConverter`

Reply:
I would say the most important would be to preserve compatibility with previous versions of Hudi - so it should be able to normally read both versions (2 level and 3 level)